### ISSUE TITLE
Autotype doesn't work when Keepass is on different virtual desktop

### ISSUE THREAD

**Issue by lschwetlick** [keepassxreboot/keepassxc#4863]
## Overview
On MacOs Catalina, KeepassXC performs the autotype in random windows when Keepass isn't running on the same virtual desktop as the window in which the autotype is triggered from.

Say I have keepassxc open and living on Desktop#2 and my browser on Desktop#1. If I press the shortcut to autotype in the browser, the focus moves to Desktop#2 where Keepass is running and gives me the dialogue of which account I want to use. When I select one, it does not move back to the  Desktop#1 before pasting the autotype. If there is another browser instance on Desktop#2 it uses that one instead, pasting my login details into the wrong page.

## Steps to Reproduce
1. Run Keepassxc on Desktop 2
2. Run Firefox on Desktop 1
3. Have Keepass set up to give you the dialogue box each time 
4. Run autotype from Desktop1/Firefox

## Expected Behavior
1. Go to Desktop 2/Keepass
2. give dialogue
3. Go to Desktop 1/ correct instance of Firefox
4. Perform autotype

## Actual Behavior
1. Go to Desktop 2/Keepass
2. give dialogue
3. Perform autotype immediately on that desktop, in a different program. If there is another firefox there, pastes into that, even if its not the correct window/tab.

## Context
KeePassXC - Version 2.5.4
Revision: dcca5aa

Qt 5.14.1
Debugging mode is disabled.

Operating system: macOS 10.15
CPU architecture: x86_64
Kernel: darwin 19.5.0

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- TouchID

Cryptographic libraries:
 libgcrypt 1.8.5

**Comment by droidmonkey** [comment]
I don't even know if this is possible, can one desktop see the windows from another desktop, let alone manipulate the keyboard on that other desktop. At this time your only choice is to run KPXC on the same Desktop as the app you want to type into.

**Comment by lschwetlick** [comment]
Just in case anyone else is having this problem, another temporary fix I found is to assign the KPXC window to all desktops (right click in the home dock->options->All Desktops). It then also shows up on all desktops which is visual clutter, but at least it doesn't paste stuff to random windows :)

**Comment by mstarke** [comment]
> I don't even know if this is possible, can one desktop see the windows from another desktop, let alone manipulate the keyboard on that other desktop. At this time your only choice is to run KPXC on the same Desktop as the app you want to type into.

The APIs you are using to read the window title allow access to all windows across virtual desktops.

**Comment by cruftex** [comment]
I have a similar or maybe even the same issue on KDE neon, which is based on Ubuntu 20.04.

Bahavior of keepassx 2.0.3 with KDE neon 18.04:
- keepassx started on virtual desktop 1
- browser started on virtual desktop 5
- auto type on active browser window in virtual desktop 5
- auto type pop-up opens above browser window on virtual desktop 5

Bahavior of keepassx 2.0.3 with KDE neon 18.04:
- keepassx started on virtual desktop 1
- browser started on virtual desktop 5
- auto type on active browser window in virtual desktop 5
- auto type pop-up opens above browser window on virtual desktop 5

Bahavior of keepassx 2.0.3 AND keepassxc 2.4.3 with KDE neon 20.04:
- keepassxc started on virtual desktop 1
- browser started on virtual desktop 5
- auto type on active browser window in virtual desktop 5
- auto type pop-up opens in virtual desktop 1, does not get focus, virtual desktop is not switched
- auto type does not work when switching to virtual desktop 1 and selecting an option from the pop up

When keepassx / keepassxc is started on the same virtual desktop as the browser, autotype works as expected.

Maybe the problem is not in keepassxc itself but a change in a library?

**Comment by droidmonkey** [comment]
Clearly a "problem" with KDE since keepassx exhibits different behavior and it obviously did not recieve any code changes between neon versions.

**Comment by ThWulf** [comment]
I observe a behavior like @cruftex on Ubuntu 18.04:

With **Keepassx 2.0.3**:
* keepassx started on virtual desktop 1
* browser started on virtual desktop 5
* auto type on active browser window in virtual desktop 5
* auto type pop-up opens above browser window on virtual desktop 5

With **Keepassxc 2.6.3**:
* keepassxc started on virtual desktop 1
* browser started on virtual desktop 5
* auto type on active browser window in virtual desktop 5
* auto type pop-up opens in virtual desktop 1, does not get focus, virtual desktop is not switched
* auto type does not work when switching to virtual desktop 1 and selecting an option from the pop up

I am a bit confused about what are the differences between keepassx and keepassxc. But if keepassxc 2.6.3. is based on keepassx 2.0.3, then something changed in the codebase. In this case @droidmonkey 's comment would not be valid in my case.

**Comment by HUGHUSR** [comment]
This is also a problem in Mojave.

2.6.x worked perfectly when using different desktops in Mojave. 
2.7 breaks autotype when on different desktops.

This REALLY needs to get fixed quickly, as most power macos users always use multiple desktops (I have 32 -- 16 across 2 displays.

THANKS!

**Comment by HUGHUSR** [comment]
Question: Why does doing autotype on macos switch to Finder before it switches to the app where the autotype is to be performed?

I ask this because in 2.7, it seems to dwell on Finder (same desktop as KP) for A LOT longer than it ever did before. Previously, you never saw the Finder window appear when doing autotype. Now, (2.7) it seems to always appear, even if for a fraction of a second.

Comment: Putting Finder in the middle has always been a PITA. I have several apps where I have to login, then enter a different password to the subfunction I wish to perform. To do the second password requires remembering to do CMD-TAB twice, else I have to go back and reselect the password field again before switching back to KP to do the autotype. I wish there was a way to take Finder out of the middle... but if that is a macos limitation, I understand. It's just annoying as heck! If you could explain why this is necessary, I would GREATLY appreciate it! 

BTW--Thanks for 2.7!! It is such a big change that I would have thought you'd called it 3.0. Fantastic work!!! We greatly appreciate all of your efforts. For such a big change, so few apparent bugs, too!! Great work!!

**Comment by HUGHUSR** [comment]
Since discovering this bug, I did some additional testing. When Finder, KP, and app are all in the same window, I have found the following 2 "changes/bugs" where 2.7 does things differently.
1) KP minimizes itself after the autotype action (shortcut key or menu) and stays minimized. In 2.6, if KP minimized itself, it restored itself so fast, you never noticed. Now, you have to click on KP in Dock to restore it. Never a problem before.
2) Autotype ONLY works when the target is browsers. I used to be able to use it to autotype in Terminal and other desktop apps, but autotype now fails -- even when KP, Finder, and app (e.g. Terminal) are all on the same virtual desktop.

**Comment by HUGHUSR** [comment]
This appears to be the same issue as bug #7722.
I would recommend they be consolidated into one bug.
They are not exact duplicates, but describe the same problem from two different perspectives.
Bug #5706 appears to be a different presentation of this same problem, but on BigSur.

**Comment by HUGHUSR** [comment]
I want to clarify that this problem is **NOT** a _New Feature_. It is a **_BUG!_**  _Please_ change its status!

What this bug report describes is an existing KP feature that is now broken.

With macOS Mojave, and all macOS versions before, you have **_always_** been able to auto-type between different desktops. That is, until KP 2.7. When I upgraded to KP 2.7, **_NOTHING_** about Mojave changed. But, auto-type between desktops works on KP 2.6.x but no longer works on 2.7. Thus, the problem **_HAS** to be in KP_ or in one of the libraries bundled into the app distribution.

The initial reporter also claims that newer versions of macOS broke on KP 2.6.x, so this is not a straight-forward issue.
Other reports (e.g., #5706) support that claim.

From my perspective, this sounds like a library issue. It appears that macOS > 10.14.x updated some library that broke KP, and now KP must be including/using some newer library reference that now (KP 2.7) breaks Mojave and earlier macOS versions. 

Please properly classify this as a _bug_, and **_please give it some priority_** as it is clearly broken in macOS, and probably broken in many X-based GUIs (i.e., KDE report, above).

**_Again, this bug requests the fixing something that used to work, but is now broken, and it IS NOT requesting something new!_**

Thanks!

Edit note:  My original wording came across as too uninformative and too negative. I apologize. Thus, I have reworded it accordingly. I hope this better clarifies the issue from a macOS Mojave viewpoint.

**Comment by phoerious** [comment]
KeePass, KeePassX, and KeePassXC are three different applications and KeePassXC never had Auto-Type on macOS to begin with.

**Comment by HUGHUSR** [comment]
> KeePass, KeePassX, and KeePassXC are three different applications and KeePassXC never hard Auto-Type on macOS to begin with.

You are right. I was thinking Linux. I have also heavily updated my original comment. I suspect that the problem is a library issue, for reasons I explain in my revised comment.
